src/d3a/models/market/two_sided.py

```diff
@@ -264,9 +277,70 @@ def match_recommendation(self, recommended_list):
                     recommended_list, index+1, trade, bid_trade
                 )
 
+    @staticmethod
+    def validate_requirements_satisfied(obj1: Union[Offer, Bid], obj2: Union[Offer, Bid]) -> None:
+        """Validate if both trade parties satisfy each other's requirements.
+
+        :raises:
+            InvalidBidOfferPair: Bid offer pair failed the validation
+        """
+        def _validate_requirements_satisfied(_obj1: Union[Offer, Bid], _obj2: Union[Offer, Bid]):
+            if _obj1.requirements:
+                for requirement in _obj1.requirements:
+                    is_satisfied = True
+                    for key, value in requirement.items():
+                        if (key == "preferred_trading_partner" and
+                                not ((hasattr(_obj2, "buyer_origin_id") and
```

Comment:
Why is it necessary to check both `buyer_origin_id` and `buyer_id`? (Same for `seller_*`)

Reply:
By `preferred_trading_partner` we could mean the seller | buyer
These, in their turns can have their offers or bids forwarded from a market to another by the IAA, in this case the IAA is the buyer|seller and thus we can't rely on this field, but the origin buyer | origin seller which are the real contributors

Reply:
But one of the two fields should always point to the original seller, right? Couldn't we just always check that field?

Reply:
One can set the market of an asset as a preferred trading agent
Let's say your house_id house_ppp, you created an offer which was forwarded to the city higher market id : ppp_city , I can set my preferred trading agent to be `ppp_city` or from `house_ppp`, so both make sense